### ISSUE TITLE
v7.12.0 update throw is missing in props validation  (react/prop-types) 

### ISSUE THREAD

**Issue by Unsfer** [yannickcr/eslint-plugin-react#2094]
After updating to v7.12.0 appeared strange validation errors.

I have this propTypes:
```js
propTypes = {
    actions: PropTypes.object.isRequired,
};
```

and this code:
```js
componentWillReceiveProps (nextProps) {
    this.props.actions.doSomething();
}

componentWillUnmount () {
    this.props.actions.doSomething();
}
```

And then I have eslint error:  'doSomething' is missing in props validation  react/prop-type
on the line in componentWillReceiveProps. But the line in componentWillUnmount is ok. 
Strange that it says "'doSomething' is missing" and not "'actions.doSomething' is missing"

I tried to change "PropTypes.object" to "PropTypes.shape", still the same.

Error disappears if I change name "componentWillReceiveProps" to something else. 
If I add propType  "doSomething: PropTypes.func.isRequired" error disappears too but it doesn't make sense.

**Comment by ljharb** [comment]
In this case, your `actions` propType should *not* be `object`, it should be `shape`, and specify that `doSomething` is a function.

**Comment by igorroch** [comment]
Similar issue when we have Ref element like this:

`    componentDidUpdate() {
        this.inputRef.focus();
    }
`

> error  'focus' is missing in props validation  react/prop-types

**Comment by ljharb** [comment]
Now *that* is a bug - accessing things chained off of `this.props` should error, but not things off of `this`.

cc @alexzherdev

**Comment by Unsfer** [comment]
As I said, I tried to change to PropTypes.shape

```js
propTypes = {
    actions: PropTypes.shape({
        doSomething: PropTypes.func.isRequired,
    }).isRequired,
};
```

Error still the same

**Comment by Studio384** [comment]
I'm having a similar issue:

```js
shouldComponentUpdate(nextProps) {
  if (this.props.search !== nextProps.search) {
    let query = nextProps.query;
    let result = nextProps.list.filter(item => {
      return (item.name.toLowerCase().includes(query.trim().toLowerCase()));
    });

    this.setState({ result });

    return true;
  }
}
```
 And am getting the following errors since 7.12

```
  ⚠  91:51  'filter' is missing in props validation       react/prop-types
  ⚠  92:37  'toLowerCase' is missing in props validation  react/prop-types
  ⚠  93:34  'toLowerCase' is missing in props validation  react/prop-types
```

**Comment by dlarin** [comment]
after update 7.12
```    componentDidUpdate (prevProps, prevState, snapshot) {
        if (!prevProps.show && this.props.show) {
            document.body.style.overflow = 'hidden'
        } else if (prevProps.show && !this.props.show) {
            document.body.style.overflow = 'auto'
        }
    }
```

```
  43:27  error  'style' is missing in props validation           react/prop-types
  43:33  error  'overflow' is missing in props validation        react/prop-types
  43:33  error  'style.overflow' is missing in props validation  react/prop-types
  45:27  error  'style' is missing in props validation           react/prop-types
  45:33  error  'overflow' is missing in props validation        react/prop-types
  45:33  error  'style.overflow' is missing in props validation  react/prop-types

✖ 6 problems (6 errors, 0 warnings)```

**Comment by devitito** [comment]
Same problem

**Comment by G100g** [comment]
Hi,
 also this code throw a error after upgrading to 7.12

```js
componentDidUpdate() {
...
        const {
            first_organization,
            second_organization,
        } = this.state;
...
```
```
155:13  error  'state.first_organization' is missing in props validation   react/prop-types
156:13  error  'state.second_organization' is missing in props validation  react/prop-types
```
                                                                                                    
Thank you for your work.

**Comment by gargroh** [comment]
Facing same issue.

**Comment by brettjthom** [comment]
Is prop validation cascading down an object new behavior?

For code like this just started failing as well. The "obj" is in props validation. ('a' is missing in props validation  react/prop-types). From what I have noticed it seems to only throw on the lifecycle functions. I use children of the object in the render and it doesn't error.

    componentWillReceiveProps(nextProps) {
        if (nextProps.obj.a!== this.props.obj.a) {
            ...
        }
    }`
`

It's also checking state it seems now. Again only noticed it on lifecycle functions
'working' is missing in props validation                 react/prop-types
`
    shouldComponentUpdate(nextProps, nextState) {
        if (nextState.working === this.state.working) {
            return false;
        }
        return true;
    }
`

**Comment by jomasti** [comment]
Looks like this is from #1946. The change in the logic for `isPropTypesUsage` to use `isInLifeCycleMethod`.

**Comment by HarikaSabbella** [comment]
When will this fix be released to NPM?

**Comment by ljharb** [comment]
[v7.12.1](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.12.1) is released.

**Comment by ajfarkas** [comment]
Stupid question: if the prop is not required, why does this rule require us to define it?
Wouldn't that make it… required?

**Comment by ljharb** [comment]
@ajfarkas "required" means that element creators can omit it at runtime; it doesn't mean that when provided it doesn't have a contract to adhere to.
